File: code/packages/elixir/branch_predictor/lib/coding_adventures/branch_predictor/two_bit.ex
defmodule CodingAdventures.BranchPredictor.TwoBit do
  @moduledoc """
  Two-bit saturating counter predictor — the classic, used in most textbooks.

  ## The four states

  The two-bit predictor improves on the one-bit predictor by adding hysteresis.
  Instead of flipping the prediction on every misprediction, it takes TWO
  consecutive mispredictions to change the predicted direction. This is achieved
  with a 2-bit saturating counter:

      STRONGLY NOT TAKEN (SNT) <-> WEAKLY NOT TAKEN (WNT) <-> WEAKLY TAKEN (WT) <-> STRONGLY TAKEN (ST)
            (00)                        (01)                       (10)                    (11)

  On "taken" outcome: move right (increment, saturate at ST)
  On "not taken" outcome: move left (decrement, saturate at SNT)

  The prediction threshold is at the midpoint:
  - States SNT, WNT -> predict NOT TAKEN
  - States WT, ST   -> predict TAKEN

  ## The DFA formalism

  Like the one-bit predictor, the two-bit predictor is formally defined as
  a DFA. The four states are {"SNT", "WNT", "WT", "ST"} and the alphabet
  is {"taken", "not_taken"}. The accepting states {"WT", "ST"} are the states
  that predict "taken".

  All transition logic is delegated to the DFA's transition table, making
  the formal specification the single source of truth.

  ## Why hysteresis works

  Consider a loop that runs 10 times:

      First invocation:
      Iter 1:  state=WNT -> predict NOT TAKEN -> actual TAKEN -> WRONG, state->WT
      Iter 2:  state=WT  -> predict TAKEN     -> actual TAKEN -> correct, state->ST
      ...
      Iter 9:  state=ST  -> predict TAKEN     -> actual TAKEN -> correct (saturated)
      Iter 10: state=ST  -> predict TAKEN     -> actual NOT TAKEN -> WRONG, state->WT

      Second invocation:
      Iter 1:  state=WT  -> predict TAKEN     -> actual TAKEN -> correct! state->ST

  Only 1 misprediction on re-entry (vs 2 for the one-bit predictor). The
  "weakly taken" state acts as a buffer.

  ## Historical usage

  - Alpha 21064: 2-bit counters with 2048 entries
  - Intel Pentium: 2-bit counters with 256 entries
  - Early ARM (ARM7): 2-bit counters with 64 entries
  - MIPS R10000: 2-bit counters as base predictor in tournament scheme

  ## Confidence mapping

  The strong/weak distinction maps naturally to confidence:
  - STRONGLY states -> 1.0 (high confidence)
  - WEAKLY states   -> 0.5 (low confidence)

  This is useful for tournament predictors that pick the most confident
  sub-predictor for each branch.
  """

  alias CodingAdventures.BranchPredictor.{Prediction, Stats}
  alias CodingAdventures.StateMachine.DFA

  # ── State name constants ──────────────────────────────────────────────────
  #
  # We use short abbreviations matching the Python implementation. These are
  # the state names in the DFA's formal definition.

  @snt "SNT"
  @wnt "WNT"
  @wt "WT"
  @st "ST"

  # ── Two-Bit DFA ──────────────────────────────────────────────────────────
  #
  # The formal state machine for the 2-bit saturating counter. Every
  # (state, input) pair maps to exactly one next state. The accepting
  # states {WT, ST} are the states that predict "taken."
  #
  # The transition table encodes the saturating counter logic:
  #   - "taken" input: move right (increment) on the state diagram
  #   - "not_taken" input: move left (decrement) on the state diagram
  #   - Saturation: ST + taken -> ST (can't go higher), SNT + not_taken -> SNT

  @two_bit_dfa (case DFA.new(
                   MapSet.new([@snt, @wnt, @wt, @st]),
                   MapSet.new(["taken", "not_taken"]),
                   %{
                     {@snt, "taken"} => @wnt,
                     {@snt, "not_taken"} => @snt,
                     {@wnt, "taken"} => @wt,
                     {@wnt, "not_taken"} => @snt,
                     {@wt, "taken"} => @st,
                     {@wt, "not_taken"} => @wnt,
                     {@st, "taken"} => @st,
                     {@st, "not_taken"} => @wt
                   },
                   @wnt,
                   MapSet.new([@wt, @st])
                 ) do
                   {:ok, dfa} -> dfa
                 end)

  defstruct table_size: 1024, initial_state: "WNT", table: %{}, stats: %Stats{}

  @type t :: %__MODULE__{
          table_size: pos_integer(),
          initial_state: String.t(),
          table: %{non_neg_integer() => String.t()},
          stats: Stats.t()
        }

  @doc """
  Create a new two-bit predictor.

  ## Options

  - `:table_size` — number of entries in the prediction table (default: 1024).
  - `:initial_state` — starting state for all counter entries (default: "WNT").
    Common choices:
    - "WNT" (Weakly Not Taken): conservative, requires 1 taken to flip
    - "WT" (Weakly Taken): optimistic, starts predicting taken

  ## Examples

      iex> p = CodingAdventures.BranchPredictor.TwoBit.new()
      iex> p.table_size
      1024
      iex> p.initial_state
      "WNT"

      iex> p = CodingAdventures.BranchPredictor.TwoBit.new(table_size: 256, initial_state: "WT")
      iex> p.initial_state
      "WT"
  """
  def new(opts \\ []) do
    table_size = Keyword.get(opts, :table_size, 1024)
    initial_state = Keyword.get(opts, :initial_state, @wnt)

    %__MODULE__{table_size: table_size, initial_state: initial_state}
  end

  @doc """
  Predict based on the 2-bit counter for this branch.

  Reads the counter state and returns taken/not-taken based on the
  threshold (WT, ST -> taken; SNT, WNT -> not-taken).

  Returns `{prediction, predictor}`.

  ## Examples

      iex> p = CodingAdventures.BranchPredictor.TwoBit.new()
      iex> {pred, _p} = CodingAdventures.BranchPredictor.TwoBit.predict(p, 0x100)
      iex> pred.predicted_taken
      false
  """
  def predict(%__MODULE__{} = predictor, pc) do
    index = index(predictor, pc)
    state = get_state(predictor, index)
    taken = predicts_taken?(state)

    confidence =
      if state in [@st, @snt] do
        1.0
      else
        0.5
      end

    {%Prediction{predicted_taken: taken, confidence: confidence}, predictor}
  end

  @doc """
  Update the 2-bit counter based on the actual outcome.

  Increments on taken, decrements on not-taken, saturating at boundaries.
  Records accuracy BEFORE updating to compare against the actual prediction.

  Returns a new predictor struct with updated table and stats.

  ## Examples

      iex> p = CodingAdventures.BranchPredictor.TwoBit.new()
      iex> p = CodingAdventures.BranchPredictor.TwoBit.update(p, 0x100, true)
      iex> {pred, _p} = CodingAdventures.BranchPredictor.TwoBit.predict(p, 0x100)
      iex> pred.predicted_taken
      true
  """
  def update(%__MODULE__{} = predictor, pc, taken, _target \\ nil) do
    index = index(predictor, pc)
    state = get_state(predictor, index)

    # Record accuracy BEFORE updating
    new_stats = Stats.record(predictor.stats, predicts_taken?(state) == taken)

    # Transition the state using the DFA
    next_state =
      if taken do
        taken_outcome(state)
      else
        not_taken_outcome(state)
      end

    %{predictor | table: Map.put(predictor.table, index, next_state), stats: new_stats}
  end

  @doc "Get prediction accuracy statistics."
  def stats(%__MODULE__{stats: stats}), do: stats

  @doc "Reset the prediction table and statistics."
  def reset(%__MODULE__{} = predictor) do
    %{predictor | table: %{}, stats: %Stats{}}
  end

  @doc """
  Compute the next state on a "taken" branch outcome.

  Delegates to the DFA transition table. This is the formal "increment"
  operation on the saturating counter.

  ## Examples

      iex> CodingAdventures.BranchPredictor.TwoBit.taken_outcome("SNT")
      "WNT"
      iex> CodingAdventures.BranchPredictor.TwoBit.taken_outcome("ST")
      "ST"
  """
  def taken_outcome(state) do
    @two_bit_dfa.transitions[{state, "taken"}]
  end

  @doc """
  Compute the next state on a "not_taken" branch outcome.

  Delegates to the DFA transition table. This is the formal "decrement"
  operation on the saturating counter.

  ## Examples

      iex> CodingAdventures.BranchPredictor.TwoBit.not_taken_outcome("ST")
      "WT"
      iex> CodingAdventures.BranchPredictor.TwoBit.not_taken_outcome("SNT")
      "SNT"
  """
  def not_taken_outcome(state) do
    @two_bit_dfa.transitions[{state, "not_taken"}]
  end

  @doc """
  Check if a state predicts "taken".

  In the DFA formalism, the accepting states are the ones that predict taken.
  States WT and ST are accepting; SNT and WNT are not.

  In hardware, this is just bit 1 of the 2-bit counter — a single wire,
  zero logic gates.

  ## Examples

      iex> CodingAdventures.BranchPredictor.TwoBit.predicts_taken?("ST")
      true
      iex> CodingAdventures.BranchPredictor.TwoBit.predicts_taken?("WT")
      true
      iex> CodingAdventures.BranchPredictor.TwoBit.predicts_taken?("WNT")
      false
      iex> CodingAdventures.BranchPredictor.TwoBit.predicts_taken?("SNT")
      false
  """
  def predicts_taken?(state) do
    MapSet.member?(@two_bit_dfa.accepting, state)
  end

  @doc """
  Return the DFA that formally defines this predictor's state transitions.

  The two-bit predictor is a 4-state DFA:
  - States: {"SNT", "WNT", "WT", "ST"}
  - Alphabet: {"taken", "not_taken"}
  - Accepting: {"WT", "ST"} (states that predict taken)
  - Initial: "WNT"

  This DFA can be visualized, verified, and traced using the state_machine
  library's tools.
  """
  def dfa, do: @two_bit_dfa

  @doc """
  Inspect the current state for a branch address (for testing/debugging).

  Returns the DFA state name (one of "SNT", "WNT", "WT", "ST") for the
  table entry that the given PC maps to.
  """
  def get_state_for_pc(%__MODULE__{} = predictor, pc) do
    get_state(predictor, index(predictor, pc))
  end

  # ── Private helpers ──────────────────────────────────────────────────────

  defp index(%__MODULE__{table_size: table_size}, pc) do
    rem(pc, table_size)
  end

  defp get_state(%__MODULE__{table: table, initial_state: initial}, index) do
    Map.get(table, index, initial)
  end
end

File: code/packages/lua/branch_predictor/src/coding_adventures/branch_predictor/two_bit.lua
-- branch_predictor/two_bit.lua — 2-bit saturating counter predictor
--
-- The two-bit predictor improves on the one-bit predictor by adding
-- HYSTERESIS. Instead of flipping the prediction on every misprediction,
-- it takes TWO consecutive mispredictions to change the predicted direction.
--
-- ## The Four States
--
--     SNT (00) — Strongly Not Taken    → predict NOT TAKEN
--     WNT (01) — Weakly Not Taken      → predict NOT TAKEN
--     WT  (10) — Weakly Taken          → predict TAKEN
--     ST  (11) — Strongly Taken        → predict TAKEN
--
-- State diagram (saturating counter):
--
--   "taken" outcome moves RIGHT (increment, saturate at ST):
--     SNT -> WNT -> WT -> ST -> ST (capped)
--
--   "not_taken" outcome moves LEFT (decrement, saturate at SNT):
--     ST -> WT -> WNT -> SNT -> SNT (capped)
--
--   Prediction threshold: WT and ST → TAKEN; SNT and WNT → NOT TAKEN
--
-- ## Why hysteresis helps
--
-- Consider a loop of 10 iterations running twice:
--
--   First invocation:
--     Iter  1: WNT → predict NOT TAKEN → actual TAKEN  → WRONG → WT
--     Iter  2: WT  → predict TAKEN     → actual TAKEN  → correct → ST
--     ...
--     Iter  9: ST  → predict TAKEN     → actual TAKEN  → correct
--     Iter 10: ST  → predict TAKEN     → actual NOT TAKEN → WRONG → WT
--
--   Second invocation:
--     Iter  1: WT  → predict TAKEN     → actual TAKEN  → correct!
--
--   Only 1 misprediction on loop re-entry (vs 2 for 1-bit predictor).
--   The "weakly taken" state acts as a buffer — a single "not taken"
--   doesn't immediately flip the prediction.
--
-- ## Historical usage
--
--   Alpha 21064: 2-bit counters with 2048 entries
--   Intel Pentium: 2-bit counters with 256 entries
--   MIPS R10000: 2-bit counters as base predictor in tournament scheme
--
-- ## Confidence
--
-- Strong states (SNT, ST) → confidence 1.0
-- Weak states  (WNT, WT)  → confidence 0.5
-- Tournament predictors use confidence to pick the best sub-predictor.

local Stats      = require("coding_adventures.branch_predictor.stats")
local Prediction = require("coding_adventures.branch_predictor.prediction")

-- State constants
local SNT = "SNT"  -- Strongly Not Taken (00)
local WNT = "WNT"  -- Weakly Not Taken   (01)
local WT  = "WT"   -- Weakly Taken       (10)
local ST  = "ST"   -- Strongly Taken     (11)

-- Transition table: state x outcome -> next state
-- Encodes the saturating counter logic.
local TAKEN_TRANSITION = {
    [SNT] = WNT,
    [WNT] = WT,
    [WT]  = ST,
    [ST]  = ST,   -- saturate
}
local NOT_TAKEN_TRANSITION = {
    [ST]  = WT,
    [WT]  = WNT,
    [WNT] = SNT,
    [SNT] = SNT,  -- saturate
}

-- States that predict TAKEN (the "accepting" states in DFA terms)
local PREDICTS_TAKEN = { [WT] = true, [ST] = true }

local TwoBit = {}
TwoBit.__index = TwoBit

-- Create a new 2-bit predictor.
--
-- Parameters:
--   table_size     (number) — number of entries (default: 1024)
--   initial_state  (string) — starting state for all entries (default: "WNT")
--     "WNT": conservative, requires 1 taken outcome to start predicting taken
--     "WT":  optimistic, starts predicting taken immediately
function TwoBit.new(table_size, initial_state)
    table_size    = table_size    or 1024
    initial_state = initial_state or WNT
    return setmetatable({
        table_size    = table_size,
        initial_state = initial_state,
        _table        = {},        -- map: index -> state string
        stats         = Stats.new(),
    }, TwoBit)
end

local function index_of(predictor, pc)
    return pc % predictor.table_size
end

local function get_state(predictor, idx)
    return predictor._table[idx] or predictor.initial_state
end

-- Predict based on the 2-bit counter for this branch.
-- Returns: prediction, predictor (predictor unchanged during predict)
function TwoBit:predict(pc)
    local idx   = index_of(self, pc)
    local state = get_state(self, idx)
    local taken = PREDICTS_TAKEN[state] == true

    local confidence
    if state == ST or state == SNT then
        confidence = 1.0   -- strong states = high confidence
    else
        confidence = 0.5   -- weak states = low confidence
    end

    return Prediction.new(taken, confidence), self
end

-- Update the 2-bit counter based on the actual outcome.
-- Increments on taken, decrements on not-taken, saturating at boundaries.
function TwoBit:update(pc, taken, _target)
    local idx   = index_of(self, pc)
    local state = get_state(self, idx)

    -- Record accuracy BEFORE updating
    local predicted = PREDICTS_TAKEN[state] == true
    local new_stats = self.stats:record(predicted == taken)

    -- Transition the saturating counter
    local next_state
    if taken then
        next_state = TAKEN_TRANSITION[state]
    else
        next_state = NOT_TAKEN_TRANSITION[state]
    end

    local new_table = {}
    for k, v in pairs(self._table) do new_table[k] = v end
    new_table[idx] = next_state

    return setmetatable({
        table_size    = self.table_size,
        initial_state = self.initial_state,
        _table        = new_table,
        stats         = new_stats,
    }, TwoBit)
end

function TwoBit:get_stats() return self.stats end

-- Inspect the current state for a branch address (for testing/debugging).
-- Returns the state string: "SNT", "WNT", "WT", or "ST".
function TwoBit:get_state_for_pc(pc)
    return get_state(self, index_of(self, pc))
end

function TwoBit:reset()
    return TwoBit.new(self.table_size, self.initial_state)
end

-- Export state constants for tests and other modules
return {
    TwoBit = TwoBit,
    SNT = SNT,
    WNT = WNT,
    WT  = WT,
    ST  = ST,
}
